### PR TITLE
Improve hierarchical term list sorting

### LINKED ISSUE #12171: Searching checklists can be problematic
The current design pattern for searching a checklist is problematic because it's possible for checked items to be hidden from view. This can lead to posts being unintentionally published into undesired categories, among other quirks. See example;

categories

We're using this current search pattern in the WooCommerce blocks we're working on where the issue is arguably more prominent. This feels like something that should be addressed in core rather than in custom blocks.

**Possible solutions**
I haven't looked into solutions in any great depth just yet but the crux is that - ideally - checked items should always be visible, even when searching. 

One option would be to employ a Chrome-inspired ⌘+F pattern which highlights results and allows you to cycle through them.

command-f

I'm posting this now to gather input before we dive further into the problem.

### PR DESCRIPTION
**Pull request by johnwatkins0**
## Description

This resolves #12171 by implementing the solution outlined in @mapk's comment here: https://github.com/WordPress/gutenberg/issues/12171#issuecomment-513031610

The first requirement, "Selected categories automatically appear at the top of the list," was already the case when the `HierarchicalTermSelector` first loads. But no further sorting would happen as the user interacted with the feature. 

This update rebuilds the tree when the selected terms change. When there's a filter term, it also rebuilds the separate filtered tree according to the requirements.

Terms are now alphabetized when not selected, and sorting is applied recursively through the tree.

## How has this been tested?

Added Jest tests that cover not just these changes but the filtering and ordering of the term list generally.

Tested locally in the browser.

## Types of changes

Non-breaking change to the internal functionality of an existing component. Also made one small CSS change that I'll note in an inline comment.

Side note: I feel the HierarchicalTermSelector component is due for a refactor. The dependency on app state and the creation of side effects within internal methods make testing the component complicated, and I think the current implementation can begin to have performance issues with long category lists. Breaking it down into three or four functional components with hooks would solve these problems. That could be done as part of this issue: https://github.com/WordPress/gutenberg/issues/17476

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js [hunk 1]
```diff
@@ -44,7 +44,7 @@ const DEFAULT_QUERY = {
 
 const MIN_TERMS_COUNT_FOR_FILTER = 8;
 
-class HierarchicalTermSelector extends Component {
+export class HierarchicalTermSelector extends Component {
```
Comment (johnwatkins0) [inline packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js]: Had to export this for testing independent of app state. Is this all right?

File: packages/editor/src/components/post-taxonomies/style.scss [hunk 2]
```diff
@@ -10,6 +10,10 @@
 	margin-bottom: 8px;
 }
 
+.editor-post-taxonomies__hierarchical-terms-choice .components-checkbox-control__input-container {
+	margin-right: 10px;
```
Comment (johnwatkins0) [inline packages/editor/src/components/post-taxonomies/style.scss]: See https://github.com/WordPress/gutenberg/issues/12171#issuecomment-576804704

**Comment by paaljoachim** [comment]
Hey John. I went ahead and added an issue I made some time ago in relation to parent - child (subcategories). I also posted the following link into the issue you linked to. https://github.com/WordPress/gutenberg/issues/15103

**Comment by annezazu** [comment]
It's been nearly a year since stale was added to this PR. I'm going to close out as a result and welcome folks to create a new PR that matches today's experience to see this work continue. Happy to reopen as well if folks would prefer for whatever reason.

**Pull request closed without merging**